internal/statesync/reactor.go

```diff
@@ -330,8 +369,10 @@ func (r *Reactor) backfill(
 			for {
 				select {
 				case height := <-queue.nextHeight():
-					r.Logger.Debug("fetching next block", "height", height)
-					lb, peer, err := r.dispatcher.LightBlock(ctxWithCancel, height)
+					peer := r.peers.Pop(ctx)
+					r.Logger.Debug("fetching next block", "height", height, "peer", peer)
+					lb, err := r.dispatcher.LightBlock(ctxWithCancel, height, peer)
```

Comment:
It looks like dispatcher here is behaving in a very similar way to how the `BlockProvider` might work. The difference is that this `Backfill` method grabs a peer from the peer list and tells the dispatcher to use the peer whereas the `BlockProvider` is constructed for one specific peer that it always uses with the dispatcher calls.

I definitely think this is an improvement since it makes clear who owns the peer list (in this case, the Reactor is responsible for maintaining the peer list). Before knowledge of the peer list was split across the reactor and the dispatcher which stretched the dispatcher abstraction pretty far.

I'm somewhat wondering why we don't just maintain a list of `BlockProviders` here as a `map[peer]BlockProvider` that we pop entries off of or delete entries from when peers go away. That way we can reuse the same code for `BlockProvider` in both places. This would mean that `dispatcher` exists soley to multiplex requests within a `BlockProvider` which I think would be a bit more consistent.